src/circuit_writer/ir.rs

```diff
@@ -0,0 +1,1119 @@
+use ark_ff::Zero;
+use circ::{
+    ir::term::{leaf_term, term, BoolNaryOp, Op, PfNaryOp, PfUnOp, Sort, Term, Value},
+    term,
+};
+use num_bigint::BigUint;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+use crate::{
+    backends::{Backend, BackendField},
+    constants::Span,
+    error::{Error, ErrorKind, Result},
+    imports::FnKind,
+    mast::Mast,
+    parser::{
+        types::{ForLoopArgument, FunctionDef, Stmt, StmtKind, TyKind},
+        Expr, ExprKind, Op2,
+    },
+    syntax::is_type,
+    type_checker::{ConstInfo, FnInfo, FullyQualified, StructInfo},
+};
+
+/// Same as [crate::var::Var], but with Term instead of ConstOrCell.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Var {
+    pub cvars: Vec<Term>,
+
+    pub span: Span,
+}
+
+impl Var {
+    pub fn new(cvars: Vec<Term>, span: Span) -> Self {
+        Self { cvars, span }
+    }
+
+    pub fn new_cvar(cvar: Term, span: Span) -> Self {
+        Self {
+            cvars: vec![cvar],
+            span,
+        }
+    }
+
+    pub fn new_var(cvar: Term, span: Span) -> Self {
+        Self {
+            cvars: vec![cvar],
+            span,
+        }
+    }
+
+    pub fn new_constant<F: BackendField>(cst: F, span: Span) -> Self {
+        let cvar = leaf_term(Op::new_const(Value::Field(cst.to_circ_field())));
+
+        Self {
+            cvars: vec![cvar],
+            span,
+        }
+    }
+
+    pub fn new_bool(b: bool, span: Span) -> Self {
+        let cvar = leaf_term(Op::new_const(Value::Bool(b)));
+
+        Self {
+            cvars: vec![cvar],
+            span,
+        }
+    }
+
+    pub fn new_constant_typ<F: BackendField>(cst_info: &ConstInfo<F>, span: Span) -> Self {
+        let ConstInfo { value, typ: _ } = cst_info;
+        let cvars = value
+            .iter()
+            .cloned()
+            .map(|f| leaf_term(Op::new_const(Value::Field(f.to_circ_field()))))
+            .collect();
+
+        Self { cvars, span }
+    }
+
+    pub fn len(&self) -> usize {
+        self.cvars.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.cvars.is_empty()
+    }
+
+    pub fn get(&self, idx: usize) -> Option<&Term> {
+        if idx < self.cvars.len() {
+            Some(&self.cvars[idx])
+        } else {
+            None
```

Comment:
yeah, it is currently handled by callers.

Reply:
Maybe we should add an issue to change that? Or do you think it's better to handle that in the callers? In any case even if you return a result it will be basically handled by callers but it would be more explicit and it will probably print a nicer error message?